content/typeclasses-syntax.md

```diff
@@ -0,0 +1,373 @@
+ 2. Otherwise, if the last parameter clause is a using (or implicit) clause, merge all parameters arising from context bounds in front of that clause, creating a single using clause.
+ 3. Otherwise, let the parameters arising from context bounds form a new using clause at the end.
+
+Rules (2) and (3) are the status quo, and match Scala 2's rules. Rule (1) is new but since context bounds so far could not be referred to, it does not apply to legacy code. Therefore, binary compatibility is maintained.
+
+**Discussion** More refined rules could be envisaged where context bounds are spread over different using clauses so that each comes as late as possible. But it would make matters more complicated and the gain in expressiveness is not clear to me.
+
+
+### 4. Context Bounds for Type Members, Deferred Givens
+
+It's not very orthogonal to allow subtype bounds for both type parameters and abstract type members, but context bounds only for type parameters. What's more, we don't even have the fallback of an explicit using clause for type members. The only alternative is to also introduce a set of abstract givens that get implemented in each subclass. This is extremely heavyweight and opaque to newcomers.
+
+**Proposal**: Allow context bounds for type members. Example:
+
+```scala
+  class Collection:
+    type Element : Ord
+```
+
+The question is how these bounds are expanded. Context bounds on type parameters
+are expanded into using clauses. But for type members this does not work, since we cannot refer to a member type of a class in a parameter type of that class. What we are after is an equivalent of using parameter clauses but represented as class members.
+
+**Proposal:**
+Introduce a new way to implement a given definition in a trait like this:
+```scala
+given T = deferred
+```
+`deferred` is a new method in the `scala.compiletime` package, which can appear only as the right hand side of a given defined in a trait. Any class implementing that trait will provide an implementation of this given. If a definition is not provided explicitly, it will be synthesized by searching for a given of type `T` in the scope of the inheriting class. Specifically, the scope in which this given will be searched is the environment of that class augmented by its parameters but not containing its members (since that would lead to recursive resolutions). If an implementation _is_ provided explicitly, it counts as an override of a concrete definition and needs an `override` modifier.
+
+Deferred givens allow a clean implementation of context bounds in traits,
+as in the following example:
+```scala
+trait Sorted:
+  type Element : Ord
+
+class SortedSet[A : Ord] extends Sorted:
+  type Element = A
+```
+The compiler expands this to the following implementation.
+```scala
+trait Sorted:
+  type Element
+  given Ord[Element] = compiletime.deferred
```

Comment:
Here is my main concern with this proposal. I think this expansion is dangerous and will create compatibility issues. Not backward incompatibility at the language, but intrinsic, synchronous incompatibility between libraries.

The core issue is that this desugaring introduces "nameless" members whose name actually matters a lot. There is no precedent for that in the language.

Consider for example
```scala
trait Babar[T]
given Babar[Int] with {}
given Babar[String] with {}

trait BaseA[A] {
  given Babar[A] = deferred
}

trait BaseB[A] {
  given Babar[A] = deferred
}

class Child extends BaseA[Int] with BaseB[String]
```
Now you need to implement two `given_Babar_A` that have nothing in common but that happen to share a generated name.

You could argue that this is already problematic with `given` definitions today, but at least they have an explicit definition. Once we automatically generate `given Ord[Element]` from `type Element : Ord`, and then `given_Ord_Element` from that, there is a multi-step, non-obvious relationship between an innocuous `type` definition and auto-generated names that clash.

Another example would be
```scala
type Foo : {package1.Bar, package2.Bar}
```
which would immediately result in clashing members. Generalize it to `type Foo : package1.Bar` in one super trait and `type Foo : package2.Bar` in another supertrait, and things becomes even more obscure.

There are so many ways that such a scheme is going to go wrong. At the level of the compilation scheme, we're approaching the amount of danger of value classes. Value classes also expose issues that arise in similar situations: where a generic in a superclass cannot be instantiated to some value classes because their bridges double-clash.

---

Since the core issue is that a generated name acquires a strong semantic meaning, I think we can fix this by disallowing anonymous things here. We could demand that deferred givens have a name, and that includes demanding them for context bounds on type members.

Even then, some issues remain: what is the visibility of the generated member? Does it match the visibility of the type member? What if the type member is `package[something]` but the implementing class is outside `something`?

Overall, this proposal introduces for the first time adding invisible members to open traits that actually matter in subtraits and subclasses. We have no precedent for that. There are a zillion issues that could arise from such a new situation, and they need to be explored much more carefully than what the current SIP text suggests.